### PR TITLE
Vivlim/sonic2 respawn spectate

### PR DESCRIPTION
**Pull request by vivlim**

### CODE REVIEW

File: sync/games/sonic2.py [hunk 1]
```diff
@@ -1,24 +1,83 @@
...
         IGameSync.__init__(self, send_data_callback, apply_data_callback)
-        self.strats.append(SimpleOnChangeStrategy([
+        self.player_position_strategy = SimpleOnChangeStrategy([
             MirroredAddress(0x10, 0xb008, 0xb048),  # pos, vel, inertia, radius
             MirroredAddress(0x1, 0xb022, 0xb062),  # status
             MirroredAddress(0x18, 0xb028, 0xb068),  # other sonic/tails-specific object fields
             MirroredAddress(0x2, 0xf604, 0xf606),  # shadow gamepad, need ROM patched to not clobber
             MirroredAddress(0x2, 0xf602, 0xf66a),  # shadow gamepad (logical)
-        ], "sonic -> tails"))
+            MirroredAddress(0x2, 0xf602, 0xf66a),  # shadow gamepad (logical)
```
Comment (lifning) [inline sync/games/sonic2.py]: why the duplicate copy?

File: sync/strategies/simple_onchange.py [hunk 2]
```diff
@@ -46,11 +47,33 @@ def write_received_data_to_memory(self, caller, received_data):
...
+                    self.received_unsynced[key] = mem
             visualization.textlog.log_text(f'Received {self.label}')
+
+    def get_last_received_value(self, offset, length, bank_switch):
+        value = self.get_value_from_received_map(self.received_unsynced, offset, length, bank_switch)
+        if value is None:
+            value = self.get_value_from_received_map(self.last_seen, offset, length, bank_switch)
+        return value
+
+    def get_value_from_received_map(self, received_map, offset, length, bank_switch):
```
Comment (lifning) [inline sync/strategies/simple_onchange.py]: should this be prefixed by _? (or will it be used by any foreign classes?)

File: sync/strategies/simple_onchange.py [hunk 3]
```diff
@@ -46,11 +47,33 @@ def write_received_data_to_memory(self, caller, received_data):
...
-                self.last_seen[key] = mem
+                if should_write:
+                    caller.write_memory(offset, mem, bank_switch)
+                    # also update last_seen so we don't send back something that we just wrote.
+                    self.last_seen[key] = mem
+                else:
+                    self.received_unsynced[key] = mem
             visualization.textlog.log_text(f'Received {self.label}')
+
+    def get_last_received_value(self, offset, length, bank_switch):
```
Comment (lifning) [inline sync/strategies/simple_onchange.py]: bank_switch=0 by default maybe?

File: sync/strategies/simple_onchange.py [hunk 4]
```diff
@@ -3,7 +3,8 @@
 
 
 class SimpleOnChangeStrategy(IStrategy):
-    last_seen = {}
+    last_seen = {}  # store synchronized received memory & keep track of memory we have sent
```
Comment (lifning) [inline sync/strategies/simple_onchange.py]: i think i'd prefer these to be defined in __init__

File: sync/games/sonic2.py (hunk 1 above)
Comment (vivlim) [reply sync/games/sonic2.py]: accident

File: sync/strategies/simple_onchange.py (hunk 2 above)
Comment (vivlim) [reply sync/strategies/simple_onchange.py]: _

File: sync/strategies/simple_onchange.py (hunk 4 above)
Comment (vivlim) [reply sync/strategies/simple_onchange.py]: sure

File: sync/strategies/simple_onchange.py (hunk 3 above)
Comment (vivlim) [reply sync/strategies/simple_onchange.py]: sure

**Review by lifning — APPROVED** [review]

**Pull request merged**